package io.github.king0929zion.zchat.data.db

import androidx.room.AutoMigration
import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.TypeConverter
import androidx.room.TypeConverters
import io.github.king0929zion.zchat.ai.core.TokenUsage
import io.github.king0929zion.zchat.data.db.dao.ConversationDAO
import io.github.king0929zion.zchat.data.db.dao.FavoriteDAO
import io.github.king0929zion.zchat.data.db.dao.GenMediaDAO
import io.github.king0929zion.zchat.data.db.dao.ManagedFileDAO
import io.github.king0929zion.zchat.data.db.dao.MemoryDAO
import io.github.king0929zion.zchat.data.db.dao.MessageNodeDAO
import io.github.king0929zion.zchat.data.db.entity.ConversationEntity
import io.github.king0929zion.zchat.data.db.entity.FavoriteEntity
import io.github.king0929zion.zchat.data.db.entity.GenMediaEntity
import io.github.king0929zion.zchat.data.db.entity.ManagedFileEntity
import io.github.king0929zion.zchat.data.db.entity.MemoryEntity
import io.github.king0929zion.zchat.data.db.entity.MessageNodeEntity
import io.github.king0929zion.zchat.data.db.migrations.Migration_16_17
import io.github.king0929zion.zchat.data.db.migrations.Migration_8_9
import io.github.king0929zion.zchat.utils.JsonInstant

@Database(
    entities = [
        ConversationEntity::class,
        MemoryEntity::class,
        GenMediaEntity::class,
        MessageNodeEntity::class,
        ManagedFileEntity::class,
        FavoriteEntity::class
    ],
    version = 17,
    autoMigrations = [
        AutoMigration(from = 1, to = 2),
        AutoMigration(from = 2, to = 3),
        AutoMigration(from = 3, to = 4),
        AutoMigration(from = 4, to = 5),
        AutoMigration(from = 5, to = 6),
        AutoMigration(from = 7, to = 8),
        AutoMigration(from = 8, to = 9, spec = Migration_8_9::class),
        AutoMigration(from = 9, to = 10),
        AutoMigration(from = 10, to = 11),
        AutoMigration(from = 12, to = 13),
        AutoMigration(from = 16, to = 17, spec = Migration_16_17::class),
    ]
)
@TypeConverters(TokenUsageConverter::class)
abstract class AppDatabase : RoomDatabase() {
    abstract fun conversationDao(): ConversationDAO

    abstract fun memoryDao(): MemoryDAO

    abstract fun genMediaDao(): GenMediaDAO

    abstract fun messageNodeDao(): MessageNodeDAO

    abstract fun managedFileDao(): ManagedFileDAO

    abstract fun favoriteDao(): FavoriteDAO
}

object TokenUsageConverter {
    @TypeConverter
    fun fromTokenUsage(usage: TokenUsage?): String {
        return JsonInstant.encodeToString(usage)
    }

    @TypeConverter
    fun toTokenUsage(usage: String): TokenUsage? {
        return JsonInstant.decodeFromString(usage)
    }
}
